### ISSUE TITLE
Iterating over a SkyCoord object is very slow

### ISSUE THREAD

**Issue by barentsen** [astropy/astropy#3323]
I was surprised to see how long it takes to iterate over a multi-valued `SkyCoord` object.

For example, it takes 19s to iterate over a SkyCoord object that contains 1000 coordinates:

``` Python
In [1]: import numpy as np
In [2]: from astropy import units as u
In [3]: from astropy.coordinates import SkyCoord
In [4]: ra = np.random.uniform(0, 360, size=1000) * u.deg
In [5]: dec = np.random.uniform(-90, 90, size=1000) * u.deg
In [6]: crd = SkyCoord(ra, dec)
In [7]: timeit crd[500]
100 loops, best of 3: 19 ms per loop
In [8]: timeit [c for c in crd]
1 loops, best of 3: 19.4 s per loop
```

In contrast, iterating over a list of 1000 SkyCoord objects is roughly 500,000x faster:

``` Python
In [9]: crd2 = [SkyCoord(ra[i], dec[i]) for i in range(len(ra))]
In [10]: %timeit crd2[500]
10000000 loops, best of 3: 38.7 ns per loop
In [11]: %timeit [c for c in crd2]
10000 loops, best of 3: 28.6 µs per loop
```

Might there be room to increase the performance of accessing a `SkyCoord` item?  Iterating over a list of coordinates seems like a common task to me.

**Comment by astrofrog** [comment]
That does sound quite slow - and we should be able to speed it up since once we have the array skycoord all the parsing etc. is done.

**Comment by embray** [comment]
Heh, I got exactly 19.4s for the first test as well.  This is definitely worth looking into.  It's no surprise that it's slower than `__getitem__` on a list, but I wouldn't be surprised if speedups are possible.

**Comment by embray** [comment]
Oh, well it's no wonder.  `SkyCoord.__getitem__` instantiates a new `SkyCoord`.  Nevertheless the second example is still faster even if you include `SkyCoord` object instantiation:
```


In [7]: %%timeit
   ...: crd = SkyCoord(ra, dec)
   ...: [c for c in crd]
   ...:
1 loops, best of 3: 19.4 s per loop

In [8]: %%timeit
   ...: crd2 = [SkyCoord(ra[i], dec[i]) for i in range(len(ra))]
   ...: [c for c in crd2]
   ...:
1 loops, best of 3: 3.16 s per loop
```

So I suspect there's room for improvement.

**Comment by taldcroft** [comment]
Yes, we could certainly use a back-door way of initializing a `SkyCoord` that just copies what's there and doesn't do any of the heavy validation that normally takes place.  It will still never be *super* fast because there is a lot of generalized machinery under the hood in the full `SkyCoord` object that is the output of `__getitem__`.

One lesson is to not do things like `crd[i]` in performant code, but instead use `crd.ra[i]` etc.

**Comment by astrofrog** [comment]
Just thinking out loud, but I wonder if we can do a similar trick here to what was done for rows in tables, i.e. when the user accesses ``SkyCoord[0].ra`` access ``SkyCoord.ra[0]``. Just an idea :)

**Comment by embray** [comment]
@astrofrog That would be doable, I think but also tricky.  It would require some sort of `SkyCoord` proxy class that can do quick lookups in the ra and dec arrays, but otherwise is not a fully initialized `SkyCoord` object until and unless any other functionality is accessed.

**Comment by astrofrog** [comment]
At the very least, even without my idea, it should be fast to initialize a ``SkyCoord`` from an existing one. The overhead should just be accessing the items in the arrays. Frames are immutable, so the existing frame can be used as-is.

**Comment by embray** [comment]
A shortcut for instantiating from `self` would definitely make a big difference here.  It looks like most of the time is wasted in `_parse_coordinate_arg`, which presumably shouldn't be necessary for this case.

**Comment by embray** [comment]
Could somebody with more coordinates expertise explain what the deal is with `FRAME_ATTR_NAMES_SET`?  It seems to me this could just be a a `TransformGraph` property in its own right that could be cached.  

Lots of time is spent in the sky_coordinate module calculating this over and over again even though it rarely changes.  What is the purpose of making it look like a module-level constant?

**Comment by embray** [comment]
I tried out my above suggestion (I left `FRAME_ATTR_NAMES_SET` as a callable but it just returns `frame_transform_graph.frame_attr_names` which is a lazy property that is cleared when the transform cache is invalidated).

Now I get:
```
In [2]: crd = SkyCoord(ra, dec)

In [3]: %timeit [c for c in crd]
1 loops, best of 3: 3.33 s per loop
```

which is barely worse than the list case.  Could still do even better but this alone made a huge difference, and probably helps in some other places as well because that `FRAME_ATTRS_NAME_SET` was being called all over the place.

**Comment by taldcroft** [comment]
@embray - IIRC `FRAME_ATTR_NAMES_SET` originally *was* a constant, but it was made dynamic in order to support dynamically-added frames within the transform graph.  Of course you are entirely correct that in practice it very rarely changes and that's kinda why I left it looking like a constant. 

In any case we should cache it.  We just need some fast key available from the transform graph object to know if it has been updated in any way.  @eteq - any suggestions?

**Comment by taldcroft** [comment]
About fast-tracking init with `self`, now I'm starting to wonder... Our nice three-layer system of SkyCoord, Frame, Representation probably means that you need to make a fast path all the way down.  Could be tricky.

**Comment by embray** [comment]
Well, I don't know about that--I don't understand the coordinates package well enough to mess with something like that.  Nevertheless, sometime tomorrow I'll post a PR with the speedups I do have (I found a few) which don't seem too intrusive and do help a bit.

**Comment by embray** [comment]
Another idea might be to implement an `__iter__` that just creates one `SkyCoord(self)` and then use that as a template for all the other returned values.  I'm not sure how well that would work in practice though since it seems like lots of math can happen on the input value between the beginning and end of `SkyCoord.__init__`.  ISTM like a good candidate for some refactoring, though I don't know what people have time for.

**Comment by eteq** [comment]
Note that #2989 is potentially relevant here too, because a lot of the time making the frames (which are inside every new `SkyCoord`) is spent doing that.  

(Will consider this a bit more soon)

**Comment by embray** [comment]
Oh, I remember that issue.  I thought it had been resolved by now but I guess not?

**Comment by cdeil** [comment]
Just now I ran into this issue where looping in Python over SkyCoords is too slow.

The use case I have is illustrated [here](http://nbviewer.ipython.org/gist/cdeil/2152b7eab2f75de138a6) ... I want to match smallish catalogs (100 to 1000 sources) using my own criteria. (At first I thought the bottleneck is the access of table columns / rows, but now I think this issue has nothing to do with Table.)

@embray Thanks for implementing #3337 to make this faster!

In the meantime ... is there a clever way to rewrite my example to get acceptable performance?
The only thing I can think of is to not use SkyCoord at all, which would be sad because it makes for really nice and clean code ...

**Comment by astrofrog** [comment]
@cdeil - if you want to iterate over coordinates at the moment, you should iterate over the arrays in skycoord, not skycoord itself. Even with #3337 it will be slow-ish because you are creating a SKyCoord every time. Much better would be:

```
c = SkyCoord(...)
for i in range(len(c)):
    ra, dec = c.ra[i], c.dec[i]
    # do stuff
```

I bet this will be faster, even after #3337.

**Comment by taldcroft** [comment]
Maybe we need some kind of lightweight iterator / index proxy.  The difference between this and what was previously mentioned with table row is that this would *not* be the default iterator so we have the freedom to deviate from the accepted coordinates API.
```
coords = SkyCoord(...)
for coord in coords.fast:  # provides a fast iterator
    ra, dec = coord.ra, coord.dec
    frame, obstime = coord.frame, coord.obstime  # just refs to parent values
    coord.full_coord  # lazy property providing the real thing (coords[index])
    coord.transform_to(FK5)  # FAIL, not a full coordinate (or maybe use full_coord.transform_to?) 

c = coords.fast[10]  # provides a fast / lightweight index access with same idea as above.
```

**Comment by astrofrog** [comment]
Thinking about this more, I think it would be good to first try and provide a fast-track initialization of frames and SkyCoord that bypasses the usual ``__init__`` stuff and sets the attributes directly without any parsing/guessing. Then the iterator can use this. This should be possible, right?

**Comment by taldcroft** [comment]
@astrofrog - when I started looking at the code I got a bit worried that we could recover really good performance.  Maybe it's possible, but with the three layers of SkyCoord, Frame, and Representation it seems like it might be tricky to really fast-track all the way down.  

To support things like iterating over pairs of coordinates you probably want microsec-type initialization while we still at the millisec level, and I'm a bit dubious about finding 2-3 orders of magnitude with fast-track.

**Comment by taldcroft** [comment]
The idea from @embray about a template object seems interesting, but I don't know how you handle the issue of being able to have arbitrary numbers of item instances around with a single per-coordinate-object template.  I.e. if you have
```
c1 = c[1]
c2 = c[2]
```
then don't you run into problems?  It seemed like the idea with the template was to have a single pre-made object where you just stuff in the correct low-level representation values as needed.  Maybe I didn't fully grok that idea.

**Comment by mhvk** [comment]
Coming late to this, and not having iiked @embray's approach to `Angle` too much, it would seem that @astrofrog is right, that a huge gain should be possible by ensuring one just takes views of the internal arrays. Obviously, one can then also handle `reshape`, etc. (as for `Time`, although that has only one layer of indirection...).

Starting at the bottom, at least for the representations, it should be trivial to have `__getitem__` not use the class initialiser but instead have a class method that simply sets the required attributes (this influences #3337; see comment there).

**Comment by embray** [comment]
Dropping this milestone on this one for now.  My PR, #3337, addresses a lot of the slowness, but still has issues.  If someone wants to pick up that work from me and finish it that would be fine too.

**Comment by woodmd** [comment]
I wanted to second the request for some more efficient way of accessing elements in a SkyCoord vector.  Being able to iterate over a list of SkyCoord objects would be much more natural than pulling values off an internal vector.  The fact that iterating over a SkyCoord object is so inefficient is also probably not going to be obvious to non-experts.  Here are some benchmarks I made with astropy v1.2.1.  You can see that there is a >1000x speed difference between the slowest and fastest methods for looping over ra and dec values in a SkyCoord vector.

Accessing SkyCoord by item:
```
c = SkyCoord(np.random.uniform(0, 360, size=200), np.random.uniform(-90, 90, size=200), unit='deg')
%timeit for i in range(len(c)): x, y = c[i].ra.deg, c[i].dec.deg
1 loop, best of 3: 9.76 s per loop
```
Accessing from internal arrays by item:
```
%timeit for i in range(len(c)): x, y = c.ra[i].deg, c.dec[i].deg
1 loop, best of 3: 445 ms per loop
```
Converting to numpy arrays:
```
%timeit for ra, dec in zip(c.ra.deg, c.dec.deg): x, y = ra, dec
100 loops, best of 3: 2.23 ms per loop
```

**Comment by mhvk** [comment]
This is a really sad difference. I started looking a bit along the chain, at at the lowest level, one can speed up getting an item from a representation by about a factor 3; see #5598.

**Comment by dmopalmer** [comment]
Pull request #5703 (discussed in #5698) increases speeds by 10x.
`%timeit crd[500]` # After #5703 2.08 ms per loop vs 21.1 ms
`%timeit [c for c in crd] ` # After #5703 2.11 s per loop vs 21.2 s

This is still slow, but it is faster.

Profiling shows that the long pole is `baseframe.py:787(_get_representation_info)` which is discussed in #2989.

**Comment by taldcroft** [comment]
Coming back to this after a while, I think that there should be something like a `fastiter` method that implements the suggestion from @embray of a template `SkyCoord`.  Since one can now easily update the representation data in place this is pretty much a snap and it will be extremely fast.

The caveat (and why it should have a distinctive name) is that this strategy is explicitly for iterating and every returned instance from the iterator will be the same object.  So doing `list(sc.fastiter())` (where `sc` is a SkyCoord) will give a list where every entry is the same.  But I think that as long as we make this abundantly clear in the docs then there are good use cases for this.  Something like the naive N**2 distance selection would be way faster than now for large N:
```
sc = SkyCoord(ra_array, dec_array, ...)
for i, c0 in enumerate(sc[:-1].fastiter()):
    for c1 in sc[i+1:].fastiter():
        compare(c0, c1)  # etc
```
Here you still need to make N new SkyCoords, but that is much better than N**2.

**Comment by mhvk** [comment]
@dmopalmer - this is very nice progress indeed!

@taldcroft - my sense would be to first speed up the current pure-python implementation; there is a lot more low-hanging fruit... (see also #5598 and leftovers of #3337)

**Comment by taldcroft** [comment]
I think this SO issue would be solved by `fastiter()`:

https://stackoverflow.com/questions/47468165/astropy-numpy-applying-function-over-coordinates-is-very-slow

This would be a very easy PR, in contrast to the overhaul of coordinate initialization that seems to be in the works.

**Comment by mhvk** [comment]
Rather than adding a method that we have to maintain indefinitely and which hopefully will be fairly useless after the speed-up efforts for 3.1, I think the easiest route here is to add an example to the documentation on how one would iterate more quickly. I think a section "Tricks for speed" would not be a bad thing to have in general.

**Comment by astrofrog** [comment]
@mhvk - https://github.com/astropy/astropy/issues/5753 😄

**Comment by astrofrog** [comment]
Rather than using a template SkyCoord, what about creating a very lightweight SkyCoordRow class that would contain a reference to a SkyCoord and a 'row' index (or more generally a view), and would then essentially slice on-the-fly when accessing attributes, and call methods with a view argument in the case of methods like transform_to (we would need to modify those methods in SkyCoord to support having a view argument). This way, each individual row will be a different object. We could even just support a subset of attributes and methods from SkyCoord to keep it simple-ish.

**Comment by taldcroft** [comment]
> which hopefully will be fairly useless after the speed-up efforts for 3.1

@mhvk - is it realistic to imagine a 2-3 order of magnitude speed-up in initialization?  That is the sort of speedup that can should be attainable with a template, and then the user has access to a bona-fide SkyCoord object.  Providing "tricks for speed" documentation is a great idea, but right now AFAIK the only way to do really fast coordinates operations is by jumping out of the coordinates framework.

About the `SkyCoordRow` idea, that is certainly interesting and would be worth investigating.  I admit that I can't immediately see how one would do this "right" and replicate the full SkyCoord API without it getting a little messy.   But that is not to say it can't be done!  I don't think we want to go down this path without intending to make the `SkyCoordRow` object fully API-compliant with a real `SkyCoord` object.

**Comment by mhvk** [comment]
@taldcroft - I got a factor of 3.5 for representations, which already did next to no work on initialization; `SkyCoord` does an *immense* amount of work, identifying the frame, taking it to pieces, and putting it all together again, while checking against all possible frame attributes. There's no need for almost all of that when one is just taking a view.  I think it can be sped up by at least a factor of 10. But I think it needs to be a bit bottom up: speed up representations, then frames, then skycoord.  In each higher class, I think one wants something like a classmethod where no checking is done at all; for `SkyCoord`, this would just set the fully populated frame and the extra attributes.

**Comment by taldcroft** [comment]
OK how about this.  Note that implementing as a `fastiter` method is just to make it easier to see what's happening and compare to the current built-in iteration.
```
    def fastiter(self, unique=True):
        if self.ndim == 0:
            raise TypeError("scalar 'SkyCoord' object is not iterable")

        comp_names = []
        self_comps = []
        for comp_name in self.representation_component_names.values():
            try:
                self_comp = getattr(self.data, comp_name)
            except AttributeError:
                pass  # For distance in UnitSphericalRepresentation
            else:
                comp_names.append('_' + comp_name)
                self_comps.append(self_comp)

        template = self[0]
        for idx in range(len(self)):
            out = copy.deepcopy(template) if unique else template
            for comp_name, self_comp in zip(comp_names, self_comps):
                setattr(out._sky_coord_frame._data, comp_name,
                        self_comp[idx])
            yield out
```
Here are the timings:
```
In [2]: cs = SkyCoord(np.ones(1000), np.ones(1000), unit='deg')

In [3]: timeit list(cs)
1 loop, best of 3: 2.92 s per loop

In [4]: timeit list(cs.fastiter())
10 loops, best of 3: 183 ms per loop

In [5]: timeit list(cs.fastiter(unique=False))
10 loops, best of 3: 39.8 ms per loop
```
It's understood that the final case is giving a list of the same object, but that is not how you would use this case.  But if you don't need unique outputs then it gives a factor of over 70 speedup.

**Comment by mhvk** [comment]
So I think we should be able to get your `unique=True` case also by cleaning up `__getitem__`. The `False` case I think is risky enough that I would rather document how to do it than to implement it.

Note that your example is not quite right in that you are not subscripting the attributes, which may or may not be a problem..

**Comment by taldcroft** [comment]
It can be even faster (factor of 20 and 500 speedups!!) by getting around Quantity a bit:
```
...
        template = self[0]
        for idx in range(len(self)):
            out = copy.deepcopy(template) if unique else template
            for comp_name, self_comp in zip(comp_names, self_comps):
                getattr(out._sky_coord_frame._data, comp_name).view(np.ndarray).__setitem__(
                    (), self_comp[idx])
            yield out

# It works (I think??)
In [2]: cs = SkyCoord(np.arange(5), np.arange(5), unit='deg')

In [3]: list(cs.fastiter())
Out[3]: 
[<SkyCoord (ICRS): (ra, dec) in deg
     ( 0.,  0.)>, <SkyCoord (ICRS): (ra, dec) in deg
     ( 1.,  1.)>, <SkyCoord (ICRS): (ra, dec) in deg
     ( 2.,  2.)>, <SkyCoord (ICRS): (ra, dec) in deg
     ( 3.,  3.)>, <SkyCoord (ICRS): (ra, dec) in deg
     ( 4.,  4.)>]

In [4]: cs = SkyCoord(np.ones(1000), np.ones(1000), unit='deg')

In [5]: timeit list(cs)
1 loop, best of 3: 2.91 s per loop

In [6]: timeit list(cs.fastiter())
10 loops, best of 3: 139 ms per loop

In [7]: timeit list(cs.fastiter(unique=False))
100 loops, best of 3: 5.74 ms per loop
```

**Comment by mhvk** [comment]
Well, we now definitely have what our efforts will be measured by! 

For 3.0, I still think this belongs in a tips-and-tricks section.

**Comment by parejkoj** [comment]
This issue came up on Facebook's Python in Astronomy recently. It looks like there are some SkyCoord-related asv benchmarks already. Would this be a good item for someone to look at for 3.1?

**Comment by astrofrog** [comment]
@parejkoj - yes - this was improved somewhat in https://github.com/astropy/astropy/commit/697dfdf9 as you can see in the drop at the end of the benchmark:

http://www.astropy.org/astropy-benchmarks/#coordinates.SkyCoordBenchmarks.time_iter_array?y-axis=log

but there's still a lot of potential for improvement. I think the plan is https://github.com/astropy/astropy/issues/7030

**Comment by mhvk** [comment]
I don't think the iteration is going to gain anything - with #6941, this is now about as fast as possible - in fact, I'll close this issue since that is the case. What #7030 will address is the general initialization (which may make iteration slightly slower again if we think the benefit of going through the initializer again outweighs the loss of speed).

**Comment by taldcroft** [comment]
> with #6941, this is now about as fast as possible

For the special case of iterating over a coordinate array where the iteration values don't need to be unique objects, there is potential for a lot more speedup (factor of ~25-ish).

@parejkoj - can you point to the facebook discussion?  I went looking in the Python in Astronomy group and didn't see anything relevant.

**Comment by mhvk** [comment]
@taldcroft - are you sure? I thought that was relative to just #6940 - with #6941 things were very similar to what you had.

**Comment by parejkoj** [comment]
@taldcroft it was a response to this post: https://www.facebook.com/groups/astropython/permalink/2030892927155387/

Looks like @bsipocz already responded.

**Comment by bsipocz** [comment]
@parejkoj - I only had #6941 in mind when I replied there (and remembered the benchmark graph, too).

**Comment by taldcroft** [comment]
@mhvk - this was the original `fastiter()` implementation that re-used the same template every time instead of doing a deepcopy.  This was a factor of [500 faster than master](https://github.com/astropy/astropy/issues/3323#issuecomment-346904684) at the time, or probably about 25 times faster than current master and 3.0.  It certainly comes with a caveat that you get the same actual object back, so `list(sky_coords)` give you a list of the same thing.  BUT there are use cases where this is OK.

**Comment by mhvk** [comment]
Ah, yes, my memory clearly was with the comparison with the unique objects. I stand corrected. Whether this further speed-up matters is of course less clear: eventually one presumably wants to *do* something with the `SkyCoord`, which likely involves a transformation that will be much slower.

Anyway, good that the aim for 3.1 is to get the regular paths as fast as we can!

**Comment by taldcroft** [comment]
On the "doing something" bit, I imagined doing something with the `ra` and `dec` attributes, not a transformation.  But just accessing a component attribute of a scalar SkyCoord is very slow, 0.36 msec on my mac.  This is 3000 times slower than accessing a direct object attribute.  Clearly SkyCoord or Frame has a lot going on there, but this looks like another avenue for performance improvement for 3.1.

**Comment by mhvk** [comment]
Indeed, that is strange.  And the more directly one goes to the real data -- avoiding the `__getattr__` overrides -- the less the problem becomes (see below). For `sc.ra`, profiling suggests most time is spent in `baseframe.get_representation_component_names` -- so this should not be too hard to solve -- but definitely worth a separate issue!
```
sc = SkyCoord(0.*u.deg, 1.*u.deg)
%timeit sc.ra
# 10000 loops, best of 3: 163 µs per loop
%timeit sc.data.lon
# 100000 loops, best of 3: 2.74 µs per loop
%timeit sc.frame.data.lon
1000000 loops, best of 3: 635 ns per loop
%timeit sc._sky_coord_frame._data.lon
1000000 loops, best of 3: 278 ns per loop
```
Conclusion: should be at least 200 times faster...

p.s. It seems clear that as long as object are immutable, we can make `__getattr__`  much faster -- indeed, one could just populate `self.__dict__`...

**Comment by taldcroft** [comment]
>  It seems clear that as long as object are immutable ..

SkyCoord and Frame objects are mutable since #5575 (astropy 1.3, Dec 2016).  The public docs show how to change these in place.  

http://astropy.readthedocs.io/en/latest/coordinates/inplace.html

**Comment by mshemuni** [comment]
Similar problem here.
I'm trying to calculate the Moon's Alt-Az coordinates for a given time at my site.
The calculation of Alt-Az angles takes too much time.
Any idea why?

Script:
```
from datetime import datetime

from time import time as tttime

from astropy.coordinates import get_moon
from astropy.coordinates import EarthLocation
from astropy.coordinates import AltAz
from astropy.coordinates import Angle
from astropy import units as U
from astropy.time import Time as tm

s = tttime()
FORMAT = "%Y-%m-%dT%H:%M:%S.%f"
date_obs = datetime.strptime("2018-07-14T08:50:32.48", FORMAT)
e = tttime()
print("Time Object created in {} seconds".format(e - s))

s = tttime()
lat = Angle("+39:46:56 degree").degree
e = tttime()
print("Angle Object created in {} seconds".format(e - s))

s = tttime()
long = Angle("+41:13:58 degree").degree
e = tttime()
print("Angle Object created in {} seconds".format(e - s))

s = tttime()
site = EarthLocation(lat=lat, lon=long, height=float(3114) * U.m)
e = tttime()
print("Site Object created in {} seconds".format(e - s))

s = tttime()
moon = get_moon(tm(date_obs, scale='utc'))
e = tttime()
print("Sky Object (Moon) created in {} seconds".format(e - s))

s = tttime()
frame_of_site = AltAz(obstime=date_obs, location=site)
e = tttime()
print("Frame Object created in {} seconds".format(e - s))

s = tttime()
object_alt_az = moon.transform_to(frame_of_site)
e = tttime()
print("Alt-Az Calculated in {} seconds".format(e - s))
```
Outpu:
```
Time Object created in 0.0 seconds
Angle Object created in 0.030791044235229492 seconds
Angle Object created in 0.0 seconds
Site Object created in 0.0009884834289550781 seconds
Sky Object (Moon) created in 0.04990100860595703 seconds
Frame Object created in 0.0 seconds
Alt-Az Calculated in 8.166500806808472 seconds
```
Python version:
`Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)] :: Anaconda, Inc. on win32`
Astropy version:
`'3.2.1'`
Numpy version:
`'1.16.5'`

**Comment by mhvk** [comment]
@mshemuni - that is really long! I just checked locally and while not as extreme as what you get, the last step does take 2 seconds.  However, just repeating the exact same script in the same session reduces it to `0.03` seconds, which suggests it is some kind of initialization issue.

Trying a bit more, I think I can localize this to reading the IERS tables (which are needed for getting accurate alt,az); if after you're import of `Time`, you add `tm.now().ut1` (which also forces the tables to be loaded), you'll see that your other statements are decently fast.

So, I think this is not an issue for `SkyCoord` itself, though I am a surprised how long the loading takes...  Looking yet a bit further, it seems clear it is because the web site for the IERS table gets contacted even if you already downloaded the file, which seems incorrect behaviour.  I think this is probably part of #9555

**Comment by mhvk** [comment]
p.s. To be a bit more specific, adding the following before anything you do makes everything fast:
```
from astropy.utils import iers
iers.IERS_Auto.open()
```

**Comment by bsipocz** [comment]
I think the refactoring  in the caching systems has solved this issue, these are the numbers I get on master when copy pasting the snippet from above, as-is.

```
Time Object created in 0.0007829666137695312 seconds
Angle Object created in 0.024201154708862305 seconds
Angle Object created in 0.00028586387634277344 seconds
Site Object created in 0.0006892681121826172 seconds
Sky Object (Moon) created in 0.04855179786682129 seconds
Frame Object created in 0.0006310939788818359 seconds
Alt-Az Calculated in 0.869938850402832 seconds
```

@mshemuni  - Could you try to update for the new release candidate and report back whether it solved the problem?  It can be pip installed: ``python -m pip install --pre -U astropy``

**Comment by mshemuni** [comment]
Exp1:
> p.s. To be a bit more specific, adding the following before anything you do makes everything fast:
> 
> ```
> from astropy.utils import iers
> iers.IERS_Auto.open()
> ```
Added two lines to my code right after import lines. Res:
`Alt-Az Calculated in 29.960163593292236 seconds`

Exp2:
Disconnected the computer from network. Res:
`Alt-Az Calculated in 0.44874143600463867 seconds`

Exp3:



> I think the refactoring in the caching systems has solved this issue, these are the numbers I get on master when copy pasting the snippet from above, as-is.
> 
> ```
> Time Object created in 0.0007829666137695312 seconds
> Angle Object created in 0.024201154708862305 seconds
> Angle Object created in 0.00028586387634277344 seconds
> Site Object created in 0.0006892681121826172 seconds
> Sky Object (Moon) created in 0.04855179786682129 seconds
> Frame Object created in 0.0006310939788818359 seconds
> Alt-Az Calculated in 0.869938850402832 seconds
> ```
> 
> @mshemuni - Could you try to update for the new release candidate and report back whether it solved the problem? It can be pip installed: `python -m pip install --pre -U astropy`

Did the update. New astropy version is `'4.0rc1'`. Res(Connected to network):
`Alt-Az Calculated in 0.7916419506072998 seconds`

The update seems to solve the problem.
Thank you @mhvk and @bsipocz for help.
